### ISSUE TITLE
[A11Y] Sidebar should disallow discovering items not in the sidebar

### ISSUE THREAD

**Issue by SchnWalter** [ampproject/amphtml#9847]
## What's the issue?

When `amp-sidebar` is open, the content outside of it should not be discover-able by screen readers.

E.g. you should not be able to explore content that is bellow the ".i-amphtml-sidebar-mask" element, which will include things like the footer links.

## How do we reproduce the issue?

1. Enable Talkback on Android (or your your favorite screen reader on your favorite device)
2. Go to the [amp-sidebar example](https://ampbyexample.com/components/amp-sidebar/)  page
3. Open the example given, using the "Toggle Sidebar" or "Open Sidebar" buttons.
4. Navigate through the elements, and once you reach the end of the sidebar content you will have the "close button" (see #5517) and then you will discover other parts of the page that should not be there until the sidebar is closed.

I'm not sure if something like `aria-hidden="true"` on the body with `aria-hidden="false"` on the sidebar would work, but something on those lines should be implemented. Maybe all the direct children of the body tag should be hidden except for the activated sidebar.

side note: You can also see this issue with the main menu on the `ampproject.org` website, but there is also an extra issue with the sub-menus being discover-able all the time and focus not being changed when a sub-menu is activated (the checkbox that controls the visibility does not lose focus), which makes the navigation painful.

## What browsers are affected?

All browsers and devices

## Which AMP version is affected?

Latest

**Comment by camelburrito** [comment]
@torch2424 Can you take care of this as a part of your project?

**Comment by torch2424** [comment]
After looking into this, I stumbled across this [SO question](https://stackoverflow.com/questions/27614714/hide-all-elements-from-screen-readers-except-one-div). Essentially, since the mask is just an element that is adjacent to the sidebar, and does not contain any of the actual content, solving this problem will prove to be an issue.

It appears, our best plan of action, would be to add role="dialog" to the amp-sidebar on open, and add aria-hidden to all direct children of body other than the sidebar. Which could be expensive depending on the size of the page, and number of children on the body.

Does this solution sound alright? Any other thoughts on the subject?

**Comment by camelburrito** [comment]
@torch2424 No this solution is not feasible. This is not just a sidebar issue , lightboxes should face the same issue too, Can you check what lightbox does ?

**Comment by torch2424** [comment]
@camelburrito Lightbox is a child element of body, that is position fixed with a high index, and covers the entire viewport. However, other elements are not children of the Lightbox, but are children on the body. 

So the same problem would be faced by Lightbox as well

**Comment by camelburrito** [comment]
So this is not just a sidebar issue. We should solve it across the board.

**Comment by SchnWalter** [comment]
Looks like the Material Design library for Angular2+ has this solved with some kind of "focus trap" functionality, and it seems to work (but I've only tested it using ChromeVox on a Linux Desktop):

https://github.com/angular/material2/blob/master/src/cdk/a11y/focus-trap.md

**Comment by aghassemi** [comment]
I am not convinced sidebar has the same semantics as a modal dialog where interaction with the background page should be prevented. The visual shim is not there to prevent interaction with the parent (unlike modal dialogs, nothing "bad" will happen if user interacts with the background) but to work as a "close" button. 

Sidebar already has a screen-reader-only close button as alternative to the shim at the end of it for closing it. Think of sidebar as a collapsible `<nav>` section which normally contains navigation links. (which is what it looks like to screen-readers at the moment)

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @camelburrito Do you have any updates?

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @torch2424 Do you have any updates?

**Comment by sparhami** [comment]
> I am not convinced sidebar has the same semantics as a modal dialog where interaction with the background page should be prevented. The visual shim is not there to prevent interaction with the parent (unlike modal dialogs, nothing "bad" will happen if user interacts with the background) but to work as a "close" button.

One problem is that many screenreaders (e.g. VoiceOver, Talkback) work by originating a click at the center coordinates of the currently selected item. When the sidebar is open and the user swipes outside and activates an item, it may actually end up "clicking" on a item in the sidebar (or closing it) instead of the action the user expects.

**Comment by sparhami** [comment]
After taking a more in depth look at this and writing some code, it looks like this is hard to solve in general. For example, if a developer has a sidebar with:

```html
<section>
<h1><span class="icon">♻</span>Recyclables</h1>
<ul>
  <li>Plastic</li>
  <li>Cans</li>
</ul>
<h1>Toggles</h1>
<input type="checkbox"></input>
</section>
```

Ideally, we would focus the `<h1>` element. But we cannot since VoiceOver will not move over if we focus it. So we cannot simply take the first element and try to focus it. We also cannot warn the developer (either as a warning or failing validation) in this situation. Sometimes we do want to find the first item (that is not the first element) and focus it. For example:

```html
<ul>
  <li>Locations</li>
  <li>Menu</li>
</ul>
```

In this case, the first `<li>` would be the right thing to focus. The `<ul>` would be the first thing we encounter, but it fails the focusability rules. You could also imagine having an image at the start of the sidebar or any other content that you would not want to get focus.

We do currently have the ability to specify what to focus. This can be done with:

```html
<amp-sidebar on="sidebarOpen:focusOnMe.focus">
    <ul>
      <li><a id="focusOnMe" href="https://google.com">This gets focused on open</a></li>
      <li><a href="#section1">Section 1</li>
    </ul>
</amp-sidebar>
```

but this has limitations too (e.g. if you render your items using an `<amp-list>`.

We also cannot be too prescriptive on the content of the sidebar. It could be just a `<nav>`, but it may not be (e.g. also has settings toggles like the first example). Another problem is that `<amp-sidebar>` is widely used, so making any changes at this point is difficult.

One idea is to perhaps make a separate accessibility validator/auditor that is optional that can make sure you specify an Element to get focus (using the `sidebarOpen` event) and that it can actually be read by VoiceOver.

**Comment by sparhami** [comment]
Looks like #12258 is a bit harder than I had hoped, so this might take a while to address as a result.

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @sparhami Do you have any updates?
